src/kimmdy/topology/topology.py

```diff
@@ -804,6 +805,118 @@ def update_parameters(self):
                 raise RuntimeError("No Parametrizer was initialized in this topology!")
             self.needs_parameterization = False
 
+    def update_partial_charges(self, recipe_steps: list[RecipeStep]) -> None:
+        """Update the topology atom partial charges based on recipe_steps.
+
+        Update rules follow a simple assignment scheme that works well with grappa.
+        If fragments are created, their partial charges are kept integers. If previously
+        broken bonds are formed again, the original partial charges are restored
+        """
+
+        def get_residue_fragments(
+            top: Topology,
+            residue: list[Atom],
+            start1: Atom,
+            start2: Atom,
+            iterations: int = 20,
+        ) -> tuple[set, set]:
+            residue_nrs = set([atom.nr for atom in residue])
+            # could remove duplicate calculations
+            fragments = [set([start1.nr]), set([start2.nr])]
+            for fragment in fragments:
+                for i in range(iterations):
+                    neighbors = set()
+                    for nr in fragment:
+                        neighbors.update(set(top.atoms[nr].bound_to_nrs))
+                    fragment.update(neighbors.intersection(residue_nrs))
+                if len(fragment) == len(residue):
+                    logger.warning(
+                        "Calculating fragments, but residue is whole! Could be break + bind reaction!"
+                    )
+                    return fragment, set()
+            if len(fragments[0]) + len(fragments[1]) != len(residue):
+                logger.warning("Residue fragments do not make up the original residue!")
+            return fragments[0], fragments[1]
+
+        # TODO: build updated list of atoms by residue for topology so that this does not need to be repeated
+        self.residues = {}
+        for atom in self.atoms.values():
+            if self.residues.get(atom.resnr) is None:
+                self.residues[atom.resnr] = []
+            self.residues[atom.resnr].append(atom)
+
+        for step in recipe_steps:
+            if isinstance(step, Break):
+                # make partial charges on either side of the break integer
+                if self.atoms[step.atom_id_1].resnr == self.atoms[step.atom_id_2].resnr:
+                    atom1 = self.atoms[step.atom_id_1]
+                    atom2 = self.atoms[step.atom_id_2]
+                    fragment1, fragment2 = get_residue_fragments(
+                        self, self.residues[atom1.resnr], atom1, atom2
+                    )
+                    if fragment2 == set():
```

Comment:
This can not be HAT atm, only intra residue reaction. To include HAT:

```suggestion
                    if len(fragment2) in (0,1):
```

Because in **inter** res HAT, one fragment would be the H